modules/optimization/src/executioners/AdjointSolve.C

```diff
@@ -0,0 +1,163 @@
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "AdjointSolve.h"
+#include "OptimizationAppTypes.h"
+
+#include "FEProblem.h"
+#include "NonlinearSystemBase.h"
+#include "NonlinearSystem.h"
+#include "NodalBCBase.h"
+
+#include "libmesh/petsc_matrix.h"
+#include "libmesh/petsc_vector.h"
+#include "petscmat.h"
+
+InputParameters
+AdjointSolve::validParams()
+{
+  InputParameters params = emptyInputParameters();
+  params.addRequiredParam<NonlinearSystemName>(
+      "forward_system", "Name of the system representing the forward problem.");
+  params.addRequiredParam<NonlinearSystemName>(
+      "adjoint_system", "Name of the system representing the adjoint problem.");
+  return params;
+}
+
+AdjointSolve::AdjointSolve(Executioner & ex)
+  : SolveObject(ex),
+    _forward_sys_num(_problem.nlSysNum(getParam<NonlinearSystemName>("forward_system"))),
+    _adjoint_sys_num(_problem.nlSysNum(getParam<NonlinearSystemName>("adjoint_system"))),
+    _nl_forward(_problem.getNonlinearSystemBase(_forward_sys_num)),
+    _nl_adjoint(_problem.getNonlinearSystemBase(_adjoint_sys_num))
+{
+  // These should never be hit, but just in case
+  if (!dynamic_cast<NonlinearSystem *>(&_nl_forward))
+    paramError("forward_system", "Forward system does not appear to be a 'NonlinearSystem'.");
+  if (!dynamic_cast<NonlinearSystem *>(&_nl_adjoint))
+    paramError("adjoint_system", "Adjoint system does not appear to be a 'NonlinearSystem'.");
+}
+
+bool
+AdjointSolve::solve()
+{
+  TIME_SECTION("execute", 1, "Executing adjoint problem", false);
+
+  if (_inner_solve && !_inner_solve->solve())
+    return false;
+
+  _problem.execute(OptimizationAppTypes::EXEC_ADJOINT_TIMESTEP_BEGIN);
+  if (!_problem.execMultiApps(OptimizationAppTypes::EXEC_ADJOINT_TIMESTEP_BEGIN))
+  {
+    _console << "MultiApps failed to converge on ADJOINT_TIMESTEP_BEGIN!" << std::endl;
+    return false;
+  }
+
+  checkIntegrity();
+
+  // Convenient references
+  // Adjoint matrix, solution, and right-hand-side
+  auto & matrix = static_cast<ImplicitSystem &>(_nl_forward.system()).get_system_matrix();
+  auto & solution = _nl_adjoint.solution();
+  auto & rhs = _nl_adjoint.getResidualNonTimeVector();
+  // Linear solver parameters
+  auto & es = _problem.es();
+  const auto tol = es.parameters.get<Real>("linear solver tolerance");
+  const auto maxits = es.parameters.get<unsigned int>("linear solver maximum iterations");
+  // Linear solver for adjoint system
+  auto & solver = *static_cast<ImplicitSystem &>(_nl_adjoint.system()).get_linear_solver();
+
+  // Assemble adjoint system by evaluating the forward Jacobian, computing the adjoint
+  // residual/source, and homogenizing nodal BCs
+  assembleAdjointSystem(matrix, solution, rhs);
+  applyNodalBCs(matrix, solution, rhs);
+
+  // Solve the adjoint system
+  solver.adjoint_solve(matrix, solution, rhs, tol, maxits);
```

Comment:
Did you have to implement the adjoint_solve method?

Reply:
No. Libmesh already has this.